### ISSUE TITLE
Using dhall with eta

### ISSUE THREAD

**Issue by jneira** [typelead/eta#704]
> Dhall is a programmable configuration language that is not Turing-complete
> You can think of Dhall as: JSON + functions + types + imports

(from https://github.com/dhall-lang/dhall-lang)

Dhall is a relatively new configuration language written in haskell and its [goals](https://github.com/dhall-lang/dhall-lang#design-philosophy) and features marry very well with langs like haskell/eta. Its main advantage over yaml f.e. is that dhall is typed and is guaranteed to terminate so you can make your config more discoverable and safe.
 * you can use it to configure programs written in other language: [haskell](https://github.com/dhall-lang/dhall-haskell), [scala](https://github.com/amarpotghan/dhall-scala) or [rust](https://github.com/nanotech/dhall-rs)
    * maybe the next can be java or groovy!
 * It can be compiled to other config formats like[ json, yaml](https://github.com/dhall-lang/dhall-json), [bash](https://github.com/dhall-lang/dhall-bash), [nix](https://github.com/dhall-lang/dhall-nix) or [cabal](https://github.com/dhall-lang/dhall-to-cabal)
   * maybe the next could be gradle

Its main drawbacks are:
  * it has a lot of dependencies
  * maybe it is too much powerful (functions and imports) for simple configurations 

The goal of this issue is track the progress and possible problems of the use of dhall ant its tools:

- [X] compilation of [dhall-haskell](https://github.com/dhall-lang/dhall-haskell/) with eta
- [X] execution with eta of all core program subcommands (dhall, *dhall repl*, dhal format, dhall hash, ...)
  - Tested repl, format, hash and lint
- [X] execution with etlas of its tools: 
  - [X] [dhall-json](https://github.com/dhall-lang/dhall-json): it needed a patch
  - [X] [dhall-text](https://github.com/dhall-lang/dhall-text), 
  - [X] [dhall-to-cabal](dhall-to-cabal)
  - [dhall-nix] (https://github.com/dhall-lang/dhall-nix)
- test performance: we will open a new one if needed
- [X] use [dhall-to-etlas](https://github.com/eta-lang/dhall-to-etlas) to *make etlas use dhall as its default configuration input*. 
  - In progress, tracked in this one: https://github.com/typelead/etlas/issues/66
- [X] create a [library](https://github.com/eta-lang/dhall-eta) to expose the dhall api and make easy the interop with java client code

**Comment by puffnfresh** [comment]
Sounds good!!

**Comment by rahulmutt** [comment]
Good stuff! Finally looked into Dhall this morning and it looks great. One cool thing to try out is to make a Dhall-to-Gradle compiler to make Gradle configuration type-safe, similar to the [Gradle Kotlin DSL](https://github.com/gradle/kotlin-dsl). It would be nice to have proper syntax highlighting for it.

Functions and imports are actually nice for configurations. It means you can provide a standard template configuration and have users modify it to their needs without having to copy/paste which can be error prone. YAML's syntax for key/value abstractions actually uses ad-hoc symbols and I'd take a `let` expression over that any day!

If we do shift to Dhall though, we should make sure we have lots of templates + examples because StackOverflow won't help people like it would with existing formats.

**Comment by jneira** [comment]
[dhall-to-cabal](https://github.com/dhall-lang/dhall-to-cabal) has been released!
the dhall expression for the cabal of the project itself:
```
let GitHub-project =
          https://raw.githubusercontent.com/ocharles/dhall-to-cabal/1.0.0/dhall/GitHub-project.dhall

in  let prelude =
          https://raw.githubusercontent.com/ocharles/dhall-to-cabal/1.0.0/dhall/prelude.dhall

in    GitHub-project
      { owner = "ocharles", repo = "example" }
    ⫽ { version = prelude.v "1.0.0"
      , library = prelude.unconditional.library
          (   prelude.defaults.Library
            ⫽ { build-depends =
                  [ { package ="base"
                    , bounds =  prelude.majorBoundVersion (prelude.v "4")
                    }
                  ]
              , exposed-modules =  [ "Hello.World" ]
              }
          )
      }
```

**Comment by rahulmutt** [comment]
Really cool :) Does Dhall have a feature to show a simplified version of the config with all the `let`'s inlined and the URLs downloaded? That will be useful for debugging configs.

**Comment by jneira** [comment]
@rahulmutt yeah, it names it "reduce to the normal form":
> The normal form is equivalent to the original program except stripped of all imports and indirection

Moreover 

> dhall-to-cabal is currently running a two-for-one deal! By installing dhall-to-cabal we'll throw in the cabal-to-dhall  executable absolutely free! cabal-to-dhall does the reverse of dhall-to-cabal - taking .cabal files and trasforming them into appropriate Dhall expressions. This can be a great way to get started with dhall-to-cabal.

(from https://github.com/dhall-lang/dhall-to-cabal/blob/master/README.md)

**Comment by rahulmutt** [comment]
That's great to hear - looks like it is going to good for practical use.

**Comment by jneira** [comment]
```
> echo True | etlas run exe:dhall
Up to date
Bool

True
```
it lives!

**Comment by jneira** [comment]
* The build has used a bunch of patches that are in [this branch of eta-hackage](https://github.com/jneira/eta-hackage/tree/dhall-patches) and a [fork of dhall-haskell](https://github.com/jneira/dhall-haskell) with some minor changes to make it compatible with eta. It used [the branch of eta with the switch from `cpphs` to `hpp` ](https://github.com/typelead/eta/tree/windows-cpp)
* Patches are incomplete: we have to implement the java ffi to double-conversion: https://github.com/jneira/eta-hackage/blob/dhall-patches/patches/double-conversion-2.0.2.0.patch
* We have to test more complex expression involving functions and http calls
* The upstream of dhall-haskell has one type application that we have to revise: https://github.com/dhall-lang/dhall-haskell/compare/master...jneira:master
* We have to try compile and execute dhall-repl component (it is commented for now)

**Comment by jneira** [comment]
* I've updated the fork of dhall to adapt last commits upstream (it replaces `text-format` package with `formatting`)
  * Also replaced the type application with a type annotation: `Data.Scientific.fromFloatDigits @Double` to `Data.Scientific.fromFloatDigits :: Double -> Scientific`
* `formatting` patch added

**Comment by jneira** [comment]
* As eta is built ghc-7.10.3/lts-6,27 and maybe we would need to build dhall with them,  i've created a branch of dhall-haskell with those constraints: https://github.com/dhall-lang/dhall-haskell/compare/master...jneira:ghc-7.10.3?expand=1
* The branch to be used *within* eta only has (at least for now) the dhall-repl commented: https://github.com/jneira/dhall-haskell/compare/ghc-7.10.3...jneira:eta?expand=1

**Comment by rahulmutt** [comment]
@jneira Awesome thanks! I think this means we can add support for something like `[package-name].dhall` and `project.dhall` for `new-build` directly inside of `etlas` by using `dhall` as a dependency with the `lts-6.27`-compatible configuration. I have no issues if you use a git dependency to your fork, but I would probably be better if the fork is under the `eta-lang` org.

Would you be interested in implementing the parser bit that will parse the `.dhall` into a proper `GenericPackageDescription`?

**Comment by jneira** [comment]
@rahulmutt Totally agree with moving the project to eta-lang. The version to use withint eta could be a patch in eta-hackage but obviously the fork to be included in eta it is not. If dhall maintainers think it is a good idea include the branch with ghc-7.10.3 compatibility in the upstream project we could use a git reference in stack.yaml. In any case i am going to move the project.

> Would you be interested in implementing the parser bit that will parse the .dhall into a proper GenericPackageDescription?

Sure!

**Comment by jneira** [comment]
It turns out that [dhall-to-cabal](https://github.com/dhall-lang/dhall-to-cabal) already does the conversion of a dhall file to a `GenericPackageDescription`: https://github.com/dhall-lang/dhall-to-cabal/blob/master/lib/DhallToCabal.hs#L550

**Comment by rahulmutt** [comment]
In that case, this line:
https://github.com/dhall-lang/dhall-to-cabal/blob/master/dhall-to-cabal.cabal#L85
Needs to be patched to `etlas-cabal`. And then you can use it as a dependency for `etlas`.

**Comment by jneira** [comment]
mmm, it seems it would need some more changes. Maybe we would need a dhall-to-etlas...

**Comment by rahulmutt** [comment]
@jneira Yes most likely that will be the case. You can create a fork and name is as `dhall-to-etlas` and add it to the eta-lang repo. Then, you can add it as a submodule for the `etlas` repo.

**Comment by rahulmutt** [comment]
So there's two things that need to be changed in the internal logic for `etlas`:
1) When `etlas` looks for `[package-name].cabal`, it should also look for `[package-name].dhall`.
2) When `etlas` looks for `cabal.project`, `cabal.project.freeze`, etc. it should also look for `etlas.dhall`, `etlas.dhall.freeze`, etc. 

How does that sound?

**Comment by jneira** [comment]
Otoh, the actual version of dhall is throwing `StackOverflowExceptions` for some simple inputs (like `True` or `False`, but no for `1`, `+1` or `1.1`). It seems the cause is the switch of dhall-haskell to `megaparsec` instead `trifecta` as parser library.
As @rahulmutt has noted, that package seems to be continuation based and it would need a patch adding a trampoline function (like aeson or binary)

**Comment by jneira** [comment]
@rahulmutt sounds good to me, i'll give a try asap

**Comment by rahulmutt** [comment]
@jneira Can you file the issue with `megaparsec` on `eta-hackage`?

**Comment by jneira** [comment]
Hi, i have a version of [dhall-to-etlas](https://github.com/eta-lang/dhall-to-etlas) that compiles!
The version has a [git reference in stack.yaml](https://github.com/eta-lang/dhall-to-etlas/blob/master/stack.yaml#L8-L10) to a commit of etlas, so it can be built outside of it. I suppose it will have to be changed to be included as a git submodule in etlas itself.
I dont fully analize the code but i dont sure if it can generate a full `GenericPackageDescription` usable in `etlas`.

**Comment by jneira** [comment]
Hi, before including dhall-to-etlas in etlas i've wanted to write a [valid dhall-to-etlas.dhall](https://github.com/jneira/dhall-to-etlas/blob/master/dhall-to-etlas.dhall). 
The result is longer than cabal version and i have to say that syntax is more noisy than cabal or yaml format. However lets, lambdas and imports are a game changer and the possibilities are very promising.
And the type checking is invaluable.

**Comment by rahulmutt** [comment]
The typechecking is certainly better than getting a `parse error` for Cabal that doesn't tell you much. The basic let abstraction is a lot more readable than yaml's alias which uses rather arbitrary symbols. I find the sequence of `let ... in`'s distracting - is there a way to have multiple let definitions in a single let like we have in Haskell/Eta?

And yes, I agree that we should tune the definitions accordingly to the common use cases for Etlas.

**Comment by jneira** [comment]
@rahulmutt yeah, it could have the lisp `let*` to make `let x = 1 ; y = x; ...` equals to `let x = 1 in let y = x...` but it seems let only can be a single expression, not sure about the reason. 
As a workaround you could define a record with the bindings, but then you have to reference the record:
```
let vars = { a = 1, b = 2, ...}
in vars.a + vars.b
```

**Comment by Gabriel439** [comment]
`let` is the way it is partly out of design and partly out of necessity

The hard constraint is that Dhall is whitespace insensitive, so Dhall has to have some separator between successive definitions and cannot rely on significant whitespace to separate successive `let` definitions.

Currently the "separator" is: `in let`, as in:

```haskell
    let x = 1
in  let y = True
in  let z = "ABC"
in  ...
```

... but could conceivably be another separator like `;` (as in Nix, which is also whitespace-insensitive).  I'm not sure if `let` itself could be the separator (i.e. `let x = 1 let y = True ...`) because I don't know if that would lead to ambiguous parses or not.

However, the real reason that I keep it that way is by design: I try to minimize alternative syntax for doing that same thing unless the productivity gain is large.  The main reasons why in roughly decreasing order of importance are:

* Users don't have to decide what is the right way to do something
    * Those sorts of minor decisions add up and drain people's willpower
    * This enthusiasm drain can outweigh the convenience if the syntactic gain is small
* The language cultivates a brand of being focused on semantics and automation rather than syntax
    * The Dhall philosophy is that it's okay to be verbose as long as you are not repetitive
    * Dhall optimizes for large and repetitive configuration files
    * Small configuration files are already well-served by existing configuration file formats
* Simplifying the language standard for ease of parallel implementations

In this specific case, the first reason is the deciding factor: removing the `in ` from `in let` only saves three characters, needlessly complicates the language and adds opportunities for confusion for people first learning the language ("Why are there two ways to do the same thing?  Are they semantically different?").  This simplification also doesn't make Dhall code easier to author or format because Dhall already has support for automatically formatting code.

I also want to add that I don't think this will be the biggest issue you will run into.  The bigger issue will be the time cost of downloading the `dhall-to-cabal` Prelude on each run of the program and being robust in the face of it being unavailable.  I'm working on adding support for using semantic integrity checks to cache these sorts of lookups but that needs to be standardized first.

**Comment by jneira** [comment]
@Gabriel439 thanks a lot for  your detailed explanation

i would agree with most of your exposition but specially with

> The language cultivates a brand of being focused on semantics and automation rather than syntax

-------------
> The bigger issue will be the time cost of downloading the dhall-to-cabal Prelude on each run of the program and being robust in the face of it being unavailable. I'm working on adding support for using semantic integrity checks to cache these sorts of lookups but that needs to be standardized first.

I was thinking that etlas could download locally dhall definitions like cached packages or other config files so that feature would be great. I suppose it is covered by https://github.com/dhall-lang/dhall-haskell/issues/162

**Comment by Gabriel439** [comment]
@jneira: I just created an issue to describe and discuss the import caching idea I mentioned: https://github.com/dhall-lang/dhall-lang/issues/129

**Comment by jneira** [comment]
Hi, actually dhall-to-etlas (like dhall-to-cabal) does not support version ranges in the `cabal-version` field, so we cant set it to `>= 1.10` without having warnings.
I've started to add support for the [actual type of etlas](https://github.com/typelead/etlas/blob/master/etlas-cabal/Distribution/Types/PackageDescription.hs#L130):
```haskell
 specVersionRaw :: Either Version VersionRange
```
with this dhall type:
```haskell
 cabal-version :
   < Left : ./Version.dhall | Right : ./VersionRange.dhall >
```
The hard part will be change haskell code to take in account the new definition :thinking:

**Comment by Gabriel439** [comment]
@jneira: If I remember correctly, @ocharles said it is because `cabal` doesn't support version ranges on `cabal-version` with an upper bound, so the version range is uniquely determined by its lower bound.  I believe in the example you gave you would specify version `1.10` and `dhall-to-cabal` would translate that to `>= 1.10` (although I could be wrong because I haven't tested this recently).

**Comment by jneira** [comment]
hi @Gabriel439, dont sure if i fully understand it but i just test dhall-to-cabal and with `cabal-version = v  ("1.10")` it generates `cabal-version: 1.10`. should it generate `>= 1.10`?

**Comment by Gabriel439** [comment]
@jneira: Then maybe I was wrong.  I'd open an issue against `dhall-to-cabal` about this if I were you

**Comment by jneira** [comment]
Btw i've opened a issue in `dhall-to-cabal` about `cabal-version` field values: https://github.com/dhall-lang/dhall-to-cabal/issues/37
In the meanwhile we could use `cabal-version = 1.12` for eta and its packages cause [ghc 7.2 shiped with that one](https://github.com/haskell/cabal/issues/5108#issuecomment-362942702).

**Comment by f-f** [comment]
Hi @jneira! Thanks for the effort in here :)

I'm trying to replicate it, and I think I'm hitting the same issue with the `StackOverflowError` when trying to evaluate some super simple input (like `[ 1 , 2 ]`) that you fixed [here](https://github.com/typelead/eta/issues/704#issuecomment-380340818).
I noticed that in the meanwhile we have `megaparsec-6.5` (which doesn't have the patch) up on eta-stackage, but even pinning the version to the patched one (`6.4.1`) doesn't fix the issue.

I'm wondering if I'm missing any relevant patches? I just cloned `dhall-haskell`, run `etlas run exe:dhall -- --explain`.

**Comment by jneira** [comment]
@f-f hi, thank you for join us and try it.
The version that i've used to build and test is a [branch of the `eta-lang` fork of `dhall-haskell`](https://github.com/eta-lang/dhall-haskell/tree/eta), but the unique diff with upstream is the dhall-repl executable is commented and afaik it does not have to do with the error.
I've reinstalled dhall-haskell with etlas again and i am able to reproduce the error. I've downloaded megaparsec-6.4.1 and now i am analyzing the stacktrace to find out the cause of the error.

**Comment by rahulmutt** [comment]
@jneira For etlas purposes, I don't think we need to worry too much about the `cabal-version` field, so using `1.12` is fine. We don't have to deal with backwards compatibility like Cabal guys do. 

Yes, the extra-libraries "hack" was to avoid creating a new field for maven-depends, but I guess it doesn't hurt to add a separate `mavenDepends` field to `BuildInfo` and update `dhall-to-etlas` accordingly - thoughts?

**Comment by jneira** [comment]
Hi, as there is a bug with last version of eta related with redirecting output to eta executables i've setup a [eta example using dhall as a lib](https://github.com/jneira/eta-test/blob/dhall/app/Main.hs) . It requires [a patch for dhall in eta-hackage](https://github.com/typelead/eta-hackage/pull/97).
The test pass:
```
C:\ws\eta\eta-test>etlas run
Up to date
1
```

Not sure if it is due to changes in dhall, eta or using dhall as a lib instead the executable...

**Comment by rahulmutt** [comment]
With a bit of debugging, I was able to isolate it to the following program:
```
import qualified Data.Text.IO

main :: IO ()
main = Data.Text.IO.getContents >>= print
```

It turns out that using the normal `getContents` from `System.IO` works fine, but `Data.Text.IO` doesn't. This may be related to the fact that we recently patched `text-1.2.3.0` which may have changed the behavior of `getContents` for `Text`.

**Comment by rahulmutt** [comment]
@jneira Fixed with https://github.com/typelead/eta/commit/964a98bd6ead37fab1fd10d091b83aa3d2775065. Turns out `hFileSize` returns `-1` if the handle is not a file handle, which we weren't doing. Surprised we didn't hit this before.

After rebuilding from scratch, I get this:
```
$ echo 1 | etlas run exe:dhall
Up to date
Natural

1
```

Can you try to reproduce the stackoverflowerror you were getting before?

**Comment by jneira** [comment]
No `StackOverflow` now in my tests!
```
C:\Users\Javier\dev\lang\eta\dhall\dhall-haskell>echo { a = 1 + 1 } | etlas run exe:dhall
Up to date
{ a : Natural }

{ a = 2 }
```

**Comment by rahulmutt** [comment]
I have been debugging this and it turns out there are yet-to-be implemented FFI calls in the `cryptonite` patch. I've fixed the error shown above with `/dev/random` locally and I'm tackling the next one involving the implementation of ChaCha.

**Comment by rahulmutt** [comment]
Updates: I've implemented Curve25519 and AES-GCM as well and the new error seems to be about a mismatch in MAC codes. 
```
Caused by: eta.runtime.exception.EtaException: HandshakeFailed (Error_Protocol ("bad record mac",True,BadRecordMac))
	at tls.network.tls.handshake.Common.handshakeFailed1(Common.hs)
	at tls.network.tls.Handshake.handshake1(Handshake.hs)
	at tls.network.tls.Handshake$sat_s8GVT.applyV(Handshake.hs)
```

See https://github.com/vincenthz/hs-tls/issues/276 for more details.

**Comment by jneira** [comment]
* Thanks to @rahulmutt  patch oh haskelline `etlas run exe:dhall -- repl` works in *nix! In windows seems to hang but we are working on this 
* Otoh i've downloaded and built `dhall-json` and the json conversion works!
```
D:\ws\eta\dhall\dhall-json>echo {a=1,b=2} | etlas run
Up to date
{"a":1,"b":2}
```
(i had to comment for now the yaml part cause it depends on `yaml` package that depends itself heavily in a c library)

**Comment by jneira** [comment]
* I've setup a project to wrap dhall-haskell and provide a friendly java api over dhall: https://github.com/eta-lang/dhall-eta
* It is a WIP but i've managed to write a simple export to be used from java:
```haskell
module Dhall.Eta where

import qualified Dhall
import qualified Data.Text as Text
import Java

foreign export java "@static org.dhall.eta.Input.bool" bool
  :: JString -> IO Bool
bool = ( Dhall.input Dhall.bool ) . Text.pack . fromJava
```
```java
package org.dhall.eta.example;

import org.dhall.eta.Input;

public class Client {
    public static void main (String[] args) {
        System.out.println("Testing dhall");
        System.out.println(Input.bool("True"));
        System.out.println(Input.bool("True && False"));
        try {
            System.out.println(Input.bool("1"));
        } catch (Exception e) {
            System.out.println(e.getLocalizedMessage());
        }
        System.out.println("The end");
    }   
}
```
Output:
```
Testing dhall
true
false

 [1;31mError [0m: Expression doesn't match annotation

- Bool
+ Natural


1 : Bool

(input):1:1

The end
```

**Comment by jneira** [comment]
We have started the integration between etlas and dhall: https://github.com/jneira/etlas/tree/dhall:
* The default configuration format for etlas will be dhall so if both dhall and cabal files are present it will choose the dhall file
* As suggested by @rahulmutt the dhall configuration file will be named `package.dhall` for all packages (like `package.yaml` of `hpack`)
* We will integrate the [caching mechanism of dhall imports](https://github.com/dhall-lang/dhall-lang/issues/129) as soon as it is available

**Comment by Gabriel439** [comment]
One thing that would help on my end is getting this integrated into `nixpkgs` (cc: @puffnfresh):

https://github.com/eta-lang/eta-nix

That would allow me to add an Eta build of Dhall to my CI so that I can catch potential Eta-related issues earlier (like adding new dependencies that might require work to port to Eta)

I tried using that repository directly but I ran into issues due to mismatch between that repository and `nixpkgs` (like missing support for the `executableHaskellDepends` field that `cabal2nix` generates)

**Comment by jneira** [comment]
@Gabriel439 that would be awesome, thank you very much
Maybe it would be useful add the build of dhall with the actual `stack-lts-6.yaml` in CI if possible, cause is the lts used by etlas/eta and it will catch some of the posssible issues (like https://github.com/dhall-lang/dhall-haskell/pull/541) .

**Comment by puffnfresh** [comment]
@Gabriel439 I've got `cabal2nix cabal://dhall` instantiated. I don't know if it will compile but I'll at least turn it into valid Nix code. I'll push in a few minutes.

**Comment by puffnfresh** [comment]
Added a chunk of overrides to get Dhall close to compiling. Got stuck on Cryptonite - it has a jar dependency and I'll have to figure out a way to supply that.

**Comment by jneira** [comment]
Actually for build master version of dhall-haskell with etlas we would need patch the new dependecy cborg-0.2.0.0: https://github.com/typelead/eta-hackage/issues/111

**Comment by rahulmutt** [comment]
Some updates on running dhall on Eta: I fixed the Bad Record MAC bug (was a bug in cryptonite patch) and the next step is to implement the zlib bits in streaming-commons.

**Comment by rahulmutt** [comment]
Ok great, https requests with the latest patches on `eta-hackage`. @jneira Can you give dhall network imports a try?

**Comment by rahulmutt** [comment]
We have patched cborg-0.2.0.0 in https://github.com/typelead/eta-hackage/commit/e400525666594657ce1d500bd554a83e45039479 - just updating since that issue wasn't linked with @jneira's previous comment.

**Comment by jneira** [comment]
Hi, [after patching the transitive dependency of cborg, serialise](https://github.com/typelead/eta-hackage/pull/114) we can build again master version of dhall. Moreover, with the recent work of @rahulmutt, http imports finally works!:
```
> echo https://raw.githubusercontent.com/dhall-lang/Prelude/35deff0d41f2bf86c42089c6ca16665537f54d75/List/replicate | etlas run exe:
dhall
Up to date
  λ(n : Natural)

→ λ(a : Type)
)
→ λ(x : a)
)
→ Natural/fold n (List a) (λ(`as` : List a) → [ x ] # `as`) ([] : List a) a)


> echo let rep=https://raw.githubusercontent.com/dhall-lang/Prelude/35deff0d41f2bf86c42089c6ca16665537f54d75/List/replicate in rep 5 Natural 1 | etlas run exe:dhall
Up to date
[ 1, 1, 1, 1, 1 ]
```

**Comment by puffnfresh** [comment]
@Gabriel439 I have `nix-build '<nixpkgs>' -A 'etaPackages.dhall'` building with my overlay.

When evaluating I get:
```
∀Peeked an empty style stack! Please report this as a bug.
```

I don't know where this problem is coming in. I'll try updating Eta and eta-hackage and see if it's still a problem.

**Comment by ocharles** [comment]
@puffnfresh do you mean when evaluating a Dhall expression you get that? It looks possibly like a bug with `prettyprinter`. I guess that `∀` is because the thing you're trying to evaluate is a forall?

**Comment by Gabriel439** [comment]
@jneira: Would a new dependency on `unix-compat` be problematic for Eta?  The reason I ask is that I'm looking for a cross-platform way to set group/other permissions on directories or files since `System.Directory` only supports setting user-permissions and `unix-compat` was the closest thing I could find that didn't involve a sub-process call to `chmod`

**Comment by jneira** [comment]
@Gabriel439 I am afraid that eta doesnt support `unix-compat` for now so we will have to patch dhall to replace those calls with java ones (or maybe patch `unix-compat` itself). Anyway, imo there is no better option (cc @rahulmut) and we already had [to patch dhall](https://github.com/typelead/eta-hackage/blob/master/patches/dhall-1.16.1.patch).

**Comment by rahulmutt** [comment]
@Gabriel439 You can go ahead and add the dependency. At worst, once the dhall integration is done, we can add an `etlas.dhall` file which doesn't contain the dependency and uses the Java standard library to perform the same function. We do eventually plan on patching `unix` and `unix-compat` once we get an idea of which functions are *actually* being used by the Hackage ecosystem. So far, we found that filesystem-related operations are the most common.

**Comment by jneira** [comment]
Although [the wrapper of dhall for java is still a wip](https://github.com/eta-lang/dhall-eta/tree/java-types) and full of bugs, i managed to implement almost all public api of `Dhall.Parser`, `Dhall.Core` and `Dhall.TypeCheck`(next will be `Dhall.Import`):

```java
package org.dhall.eta.example;

import org.dhall.core.*;
import org.dhall.core.expr.ExprIntegerShow;
import org.dhall.parser.*;
import org.dhall.parser.error.*;
import org.dhall.typecheck.TypeError;

import org.dhall.eta.Parser;
import org.dhall.eta.TypeCheck;
import org.dhall.eta.Core;

import org.haskell.types.*;

public class Client {

    public static void main (String[] args) {
        Either<ParseError,Expr<Src,Import>> parsed = Parser.exprFromText("example","1");
        System.out.println("Parsing \"1\":");
        System.out.println(parsed);
        System.out.println("-----------------------------");

        Either<ParseError,Expr<Src,Import>> eParsedFun = 
        		Parser.exprFromText("example","λ (t : Text) -> 1");
        System.out.println("Parsing \"λ (t : Text) -> 1\":");
        System.out.println(eParsedFun);
        System.out.println("-----------------------------");
        
        Either.Matcher<ParseError, Expr<Src,Import>, Expr<Src,Import>> matcher = 
        		new Either.Matcher<>(any -> null);
        matcher.Right(r -> r.getValue());
        Expr<Src,Import> parsedFun =  matcher.match(eParsedFun);
        System.out.println("Pretty printing parsed fun");
        System.out.println(Core.pretty(parsedFun));
        System.out.println("-----------------------------");
        
        Expr<Unit,Import> norm = Core.normalize(parsedFun);
        System.out.println("Normalize fun");
        System.out.println(norm);
        System.out.println("-----------------------------");
        
        Expr<Unit,Import> alphaNorm = Core.alphaNormalize(norm);
        System.out.println("Alpha normalize fun");
        System.out.println(alphaNorm);
        System.out.println("-----------------------------");
        
        Expr<Src,Void> importedExpr = new ExprIntegerShow<>();
        Either<TypeError<Src, Void>,Expr<Src,Void>> checked = TypeCheck.typeOf(importedExpr);
        System.out.println("Type of ExprIntegerShow");
        System.out.println(checked);
        System.out.println("-----------------------------");
        
        System.out.println("The end");
    }
    
}
```
with this output:
```
Parsing "1":
Right [ExprNote [annotation=Src{begin=SourcePos{path=example,line=Pos[1],column=Pos[1]},end=SourcePos{path=example,line=Pos[1],column=Pos[2]},text="1"}, subexpr=ExprNatural [Natural [1]]]]
-----------------------------
Parsing "λ (t : Text) -> 1":
Right [ExprNote [annotation=Src{begin=SourcePos{path=example,line=Pos[1],column=Pos[1]},end=SourcePos{path=example,line=Pos[1],column=Pos[18]},text="λ (t : Text) -> 1"}, subexpr=ExprLam [name=t, type=ExprNote [annotation=Src{begin=SourcePos{path=example,line=Pos[1],column=Pos[8]},end=SourcePos{path=example,line=Pos[1],column=Pos[12]},text="Text"}, subexpr=ExprText], body=ExprNote [annotation=Src{begin=SourcePos{path=example,line=Pos[1],column=Pos[17]},end=SourcePos{path=example,line=Pos[1],column=Pos[18]},text="1"}, subexpr=ExprNatural [Natural [1]]]]]]
-----------------------------
Pretty printing parsed fun
λ(t : Text) → 1
-----------------------------
Normalize fun
ExprLam [name=t, type=ExprText, body=ExprNatural [Natural [1]]]
-----------------------------
Alpha normalize fun
ExprLam [name=_, type=ExprText, body=ExprNatural [Natural [1]]]
-----------------------------
Type of ExprIntegerShow
Right [ExprPi [name=_, kind=ExprInteger, body=ExprText]]
-----------------------------
The end
```

**Comment by jneira** [comment]
Hi, more advances in `dhall-eta`, it has [more complete wrap over dhall main module](https://github.com/eta-lang/dhall-eta/blob/java-types/src/main/eta/Dhall/Eta.hs), including a converter from dhall to classes. Suggestions about the api are welcome!

```java
public class Client {

    public static void main (String[] args) {
        
        System.out.println("Testing dhall main module");
        
        Type<Boolean> tyBool = Types.bool();
        Boolean bool = Input.type(tyBool, "True");
        System.out.println(bool);
        
        Type<Optional<String>> optStrTy=Types.optional(Types.str());
        Optional<String> optStrIn = Input.type(optStrTy, "Some \"hello\"");
        System.out.println(optStrIn);
        
        Type<List<Natural>> nsTy = Types.list(Types.natural());
        List<Natural> ns = Input.type(nsTy, "[1, 2, 3]");
        System.out.println(ns);
        
        Project p = Input.type(Types.record(new ProjectType()),
        		"{ name = \"dhall\", description = \"desc\", stars = 123 }");
        System.out.println(p);
        System.out.println("The end");
    }
    
    public static class Project {
    	private String name;
    	private String description;
        private Natural stars;
	public Project(String name, String description, Natural stars) {
		super();
		this.name = name;
		this.description = description;
		this.stars = stars;
	}
	// getters and toString boilerplate
    }
    
    public static class ProjectType implements RecordType<Project> {
		
	public Map<String,Type<? extends Object>> getFieldsTypes() {
		LinkedHashMap<String,Type<? extends Object>> fields= new LinkedHashMap<>();
		fields.put("name", Types.str());
		fields.put("description", Types.str());
		fields.put("stars", Types.natural());
		return fields;
	}
		
	public Project fromFieldsValues(Map<String,Object> m) {
		return new Project((String)m.get("name") ,
				   (String)m.get("descripcion"), 
				   (Natural)m.get("stars"));
	}
   }
}
```
Output:
```
Testing dhall main module
true
Optional[hello]
[Natural [1], Natural [2], Natural [3]]
Project [name=dhall, description=null, stars=Natural [123]]
The end
```

**Comment by jneira** [comment]
We already have released the eta lib to expose a java friendly api! 
https://github.com/eta-lang/dhall-eta

**Comment by ocharles** [comment]
Super stuff @jneira! Is this issue ready to be closed?

**Comment by jneira** [comment]
Well, only left a more complete etlas integration but we could close this one, if nobody disagrees
